backends/tfhe-cuda-backend/cuda/src/integer/scalar_shifts.cuh

```diff
@@ -107,4 +109,127 @@ __host__ void host_integer_radix_scalar_shift_kb_inplace(
+    radix_blocks_rotate_left<<<num_blocks, 256, 0, stream->stream>>>(
+        rotated_buffer, lwe_array, rotations, num_blocks, big_lwe_size);
+    cuda_memcpy_async_gpu_to_gpu(lwe_array, rotated_buffer,
+                                 num_blocks * big_lwe_size_bytes, stream);
+    cuda_synchronize_stream(stream);
+
+    if (num_bits_in_block == 1) {
+      // if there is only 1 bit in the msg part, it means shift_within block is
+      // 0 thus only rotations is required.
+
+      // We still need to pad with the value of the sign bit.
+      // And here since a block only has 1 bit of message
+      // we can optimize things by not doing the pbs to extract this sign bit
+
+      Torus *block_src =
+          rotated_buffer + (num_blocks - rotations - 1) * big_lwe_size;
+      Torus *block_dest =
+          rotated_buffer + (num_blocks - rotations) * big_lwe_size;
+      for (uint i = 0; i < num_blocks; i++) {
+        cuda_memcpy_async_gpu_to_gpu(block_dest, block_src, big_lwe_size_bytes,
+                                     stream);
+        block_dest += big_lwe_size;
+      }
+      return;
+    }
+
+    // In the arithmetic shift case we have to pad with the value of the sign
+    // bit. This creates the need for a different shifting lut than in the
+    // logical shift case. We also need another PBS to create the padding block.
+    Torus *last_block = lwe_array + (num_blocks - rotations - 1) * big_lwe_size;
+    cuda_memcpy_async_gpu_to_gpu(last_block_copy,
+                                 rotated_buffer + (num_blocks - rotations - 1) *
+                                                      big_lwe_size,
+                                 big_lwe_size_bytes, stream);
+    auto partial_current_blocks = lwe_array;
+    auto partial_next_blocks = &rotated_buffer[big_lwe_size];
+    size_t partial_block_count = num_blocks - rotations;
+    if (shift_within_block != 0 && rotations != num_blocks) {
+      integer_radix_apply_bivariate_lookup_table_kb<Torus>(
+          stream, partial_current_blocks, partial_current_blocks,
+          partial_next_blocks, bsk, ksk, partial_block_count, lut_bivariate);
+    }
+    cuda_synchronize_stream(stream);
+#pragma omp parallel sections
+    {
+      // All sections may be executed in parallel
+#pragma omp section
+      {
+        integer_radix_apply_univariate_lookup_table_kb(
+            mem->local_stream_1, padding_block, last_block_copy, bsk, ksk, 1,
+            lut_univariate_padding_block);
+        // Replace blocks 'pulled' from the left with the correct padding block
+        for (uint i = 0; i < rotations; i++) {
+          cuda_memcpy_async_gpu_to_gpu(
+              lwe_array + (num_blocks - rotations + i) * big_lwe_size,
+              padding_block, big_lwe_size_bytes, mem->local_stream_1);
+        }
+      }
+#pragma omp section
+      {
+        if (shift_within_block != 0 && rotations != num_blocks) {
+          integer_radix_apply_univariate_lookup_table_kb(
```

Comment:
Maybe it would be better to move this conditional outside the `omp` clause so that if it is not satisfied we can avoid the stream synchronization at the cost of maybe increasing redundant code.